riscv/triggers.cc

```diff
@@ -17,6 +17,94 @@ action_t trigger_t::legalize_action(reg_t val) const noexcept {
 }
 
+reg_t trigger_t::tdata3_read(const processor_t * const proc) const noexcept {
+  auto xlen = proc->get_xlen();
+  reg_t tdata3 = 0;
+  if (xlen == 32) {
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_MHVALUE, mhvalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_MHSELECT, mhselect);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_SBYTEMASK, sbytemask);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_SVALUE, svalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA32_SSELECT, sselect);
+  } else {
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_MHVALUE, mhvalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_MHSELECT, mhselect);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_SBYTEMASK, sbytemask);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_SVALUE, svalue);
+    tdata3 = set_field(tdata3, CSR_TEXTRA64_SSELECT, sselect);
+  }
+  return tdata3;
+}
+
+void trigger_t::tdata3_write(processor_t * const proc, const reg_t val) noexcept {
+  auto xlen = proc->get_xlen();
+  if (xlen == 32) {
+    mhvalue = get_field(val, CSR_TEXTRA32_MHVALUE);
+    mhselect = get_field(val, CSR_TEXTRA32_MHSELECT);
+    sbytemask = get_field(val, CSR_TEXTRA32_SBYTEMASK);
+    svalue = get_field(val, CSR_TEXTRA32_SVALUE);
+    sselect = get_field(val, CSR_TEXTRA32_SSELECT);
+  } else {
+    mhvalue = get_field(val, CSR_TEXTRA64_MHVALUE);
+    mhselect = get_field(val, CSR_TEXTRA64_MHSELECT);
+    sbytemask = get_field(val, CSR_TEXTRA64_SBYTEMASK);
+    svalue = get_field(val, CSR_TEXTRA64_SVALUE);
+    sselect = get_field(val, CSR_TEXTRA64_SSELECT);
+  }
+  if (!proc->extension_enabled('H'))
+    mhselect = mhselect & MHSELECT_MCONTEXT; // the only legal value is 4, otherwise 0 (ignore)
+  if ((mhselect & MHSELECT_RESERVED) == MHSELECT_RESERVED) // 3 and 7 are reserved for mhselect
+    mhselect = 0;
+  if (!proc->extension_enabled('S'))
+    svalue = 0;
+  if (!proc->extension_enabled('S') || sselect > SSELECT_MAXVAL)
+    sselect = SSELECT_IGNORE;
+}
+
+bool trigger_t::textra_match(processor_t * const proc) const noexcept
+{
+  auto xlen = proc->get_xlen();
+  auto hsxlen = proc->get_xlen(); // use xlen since no hsxlen
+  state_t * const state = proc->get_state();
+
+  assert(sselect <= SSELECT_MAXVAL);
+  if (sselect == SSELECT_SCONTEXT) {
+    reg_t mask = (reg_t(1) << ((xlen == 32) ? CSR_TEXTRA32_SVALUE_LENGTH : CSR_TEXTRA64_SVALUE_LENGTH)) - 1;
+    assert(CSR_TEXTRA32_SBYTEMASK_LENGTH < CSR_TEXTRA64_SBYTEMASK_LENGTH);
+    for (int i = 0; i < CSR_TEXTRA64_SBYTEMASK_LENGTH; i++)
+      if (sbytemask & (1 << i))
+        mask &= 0xff << (i * 8);
+    if ((state->csrmap[CSR_SCONTEXT]->read() ^ svalue) & mask)
+      return false;
+  } else if (sselect == SSELECT_ASID) {
+    reg_t satp = state->v ? state->vsatp->read() : state->satp->read();
+    reg_t satp_asid = xlen == 32 ? SATP32_ASID : SATP64_ASID;
+    reg_t asid = get_field(satp,  satp_asid);
+    reg_t asidmax = xlen == 32 ? 9 : 16;
```

Comment:
The spec uses the term ```ASIDMAX``` instead of ```ASIDLEN```. I believe that means it is a constant value instead of an implementation-defined value.
![image](https://user-images.githubusercontent.com/39526191/205810077-5d0340ea-40f1-4889-9b83-3db716675a97.png)
I support your point that there should be a macro ```#define ASIDMAX(XLEN) (XLEN==32?9:16)```. However, I need help deciding the file to define the macro.

Reply:
That kind of thing usually goes in `encoding.h`, which is created from [another repo](https://github.com/riscv/riscv-opcodes).

My question, though, is about implementations where ASIDLEN < ASIDMAX. Will this code still work as-is?

Reply:
The debug spec does not consider the ASIDLEN value here. For ASIDLEN < ASIDMAX, the higher bits are 0. The trigger compares the value as-is and can only fire on setting the corresponding bits 0.
![image](https://user-images.githubusercontent.com/39526191/206083188-98ec8ead-98a4-4221-a4e2-7267e7ba5968.png)
Does this behavior make sense to you?